substrate/client/network/sync/src/engine.rs

```diff
@@ -713,11 +713,36 @@ where
 			self.is_major_syncing
 				.store(self.chain_sync.status().state.is_major_syncing(), Ordering::Relaxed);
 
+			// Perform actions requested by `ChainSync` in during `select!`.
+			self.perform_chain_sync_actions();
+
 			// Send outbound requests on `ChanSync`'s behalf.
 			self.send_chain_sync_requests();
```

Comment:
Is it necessary to have both of these functions? Conceptually block requests should be sent when some action is performed for `ChainSync` (peer connected, import queue drained) which should result in further actions.

Reply:
May be it's possible to refactor the code in a way so new requests are generated only in response to some event, but right now the are generated "on their own" by going over the peers in `ChainSync` and checking if we can perform a request to each of them.

Reply:
@altonen do you think we can merge this PR? It's hard to say if it's possible to get rid of a separate stage of generating requests in `ChainSync`, but it should probably go as a separate change anyway.

Reply:
Sorry, I didn't mean it would be done in this PR either way but something to look into for sure. Might be slightly nastier than it looks on the surface because `peer_connected()/block_imported()` would likely have to call `send_requests()` or something like that. But please take a look at some point if we could simplify this further.